### PR TITLE
Add removing dangling results in MultiSubGraphOp transformation

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Sometimes outputs in `MultiSubGraphOp` are unused, but they are not deleted, and they keep parts of bodies from deleting, which could result in inputs deleted and can help reduce number of operations in the model.*

### Tickets:
 - *102884*

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp [hunk 1]
```diff
@@ -97,8 +146,47 @@ ov::pass::RemoveMultiSubGraphOpDanglingParams::RemoveMultiSubGraphOpDanglingPara
...
+        if (!outputs_to_remove.empty()) {
+            // we need to reconstruct operation with new number of outputs, we cannot reduce number of outputs of
+            // existing op
+            std::shared_ptr<ov::op::util::MultiSubGraphOp> new_op;
+            if (if_op) {
+                new_op = std::make_shared<opset10::If>(if_op->input_value(0));
+                new_op->set_arguments(multi_subgraph_op->input_values());
+            } else if (loop_op) {
+                auto new_loop_op = std::make_shared<opset10::Loop>(loop_op->input_value(0), loop_op->input_value(1));
+                new_loop_op->set_special_body_ports(loop_op->get_special_body_ports());
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp:158]: are you sure if the special port won't change the id after removing outputs?

File: src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp [hunk 2]
```diff
@@ -97,8 +146,47 @@ ov::pass::RemoveMultiSubGraphOpDanglingParams::RemoveMultiSubGraphOpDanglingPara
...
+                new_op->set_input_descriptions(body_idx, multi_subgraph_op->get_input_descriptions(body_idx));
+                new_op->set_output_descriptions(body_idx, new_op_out_desc.at(body_idx));
+            }
+            size_t removed_outs_counter = 0;
+            new_op->set_output_size(multi_subgraph_op->get_output_size() - outputs_to_remove.size());
+            for (size_t out_idx = 0; out_idx < multi_subgraph_op->get_output_size(); ++out_idx) {
+                if (outputs_to_remove.find(out_idx) != outputs_to_remove.end()) {
+                    ++removed_outs_counter;
+                    // Need to go through all output descriptors to reflect deleted output
+                    for (int body_idx = 0; static_cast<size_t>(body_idx) < subgraphs_size; ++body_idx) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp:177]: I believe this is the same as for inputs, could we re-use this code?

File: src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp [hunk 3]
```diff
@@ -4,29 +4,77 @@
...
-    ov::matcher_pass_callback callback = [=](pattern::Matcher& m) {
-        auto multi_subgraph_op = std::dynamic_pointer_cast<op::util::MultiSubGraphOp>(m.get_match_root());
-        if (multi_subgraph_op == nullptr) {
-            return false;
+bool ov::pass::RemoveMultiSubGraphOpDanglingParamsResults::run_on_model(const std::shared_ptr<ov::Model>& m) {
+    RUN_ON_MODEL_SCOPE(RemoveMultiSubGraphOpDanglingParamsResults);
+    bool is_changed = false;
+    auto ops = m->get_ordered_ops();
+    // Going in reverse order
+    for (auto it = ops.rbegin(); it != ops.rend(); ++it) {
```
[the remark this answers is not in the document]
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp:19]: Not possible, need reverse order iteration.

File: src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp (hunk 1 above)
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp:158]: Agree, need to ensure that it gets changed if necessary

File: src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp:19]: right

File: src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp (hunk 2 above)
Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp:177]: Unfortunately I don't see a way to reuse the lambda function effectively, I would create a new lambda for that

**Pull request merged**